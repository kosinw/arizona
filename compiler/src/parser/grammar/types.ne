_Type ->
      NativeType    {% id %}
    | Identifier    {% id %}

Type -> 
    _Type       {% id %}