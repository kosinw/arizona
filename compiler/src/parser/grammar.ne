# Grammar for Arizona programming langauge

@preprocessor typescript 
@builtin "whitespace.ne"

@{%
    import * as moo from 'moo';
    import { makeLexer } from '.';
    import factory, { flatten, nth, nil, compose, drop } from './syntax';

    const lexer = makeLexer();
    const syntax = factory(lexer);
%}

@lexer lexer

Expression -> AssignmentExpression  {% id %}

AssignmentExpression -> Subscript {% id %}

Call ->
    Subscript _ LP _ ArgumentList _ RP          {% compose(syntax.call, flatten) %}
    | Subscript _ LP _ RP                       {% syntax.call %}
    | Subscript                                 {% id %}

ArgumentList ->
    Expression                                  {% id %}
    | NativeType                                {% id %}
    | Expression _ COMMA _ ArgumentList         {% flatten %}

Subscript ->
    Access LSB _ Expression _ RSB Subscript     {% syntax.subscript %}
    | Access LSB _ Expression _ RSB             {% syntax.subscript %}
    | Access                                    {% id %}

Access ->
    Access DOT Identifier           {% compose(syntax.access, drop) %}
    | NativeType DOT Identifier     {% compose(syntax.access, drop) %}
    | Grouping                      {% id %}

Grouping ->
    LP _ Expression _ RP            {% nth(2) %}
    | Atom                          {% id %}

Atom -> 
    Identifier                  {% id %}
    | StringLiteral             {% id %}
    | CharLiteral               {% id %}
    | FloatLiteral              {% id %}
    | IntegerLiteral            {% id %}

NativeType      -> %type            {% syntax.type %}
Identifier      -> %identifier      {% syntax.identifier %}
FloatLiteral    -> %float           {% syntax.float %}
IntegerLiteral  -> %integer         {% syntax.integer %}
StringLiteral   -> %string          {% syntax.string %}
CharLiteral     -> %char            {% syntax.char %}
BoolLiteral     -> %bool            {% syntax.bool %}

# Punctuators
LSB       -> "["     {% nil %}
RSB       -> "]"     {% nil %}
LP        -> "("     {% nil %}
RP        -> ")"     {% nil %}
DOT       -> "."     {% nil %}
COMMA     -> ","     {% nil %}

# Keywords
FN        -> "fn"                   {% nil %}
LET       -> "let"                  {% nil %}
CONST     -> "const"                {% nil %}
EXPORT    -> "export"               {% nil %}
IMPORT    -> "import"               {% nil %}
FROM      -> "from"                 {% nil %}
RETURN    -> "return"               {% nil %}
TYPE      -> "type"                 {% nil %}
IF        -> "if"                   {% nil %}
ELSE      -> "else"                 {% nil %} 
FOR       -> "for"                  {% nil %}
WHILE     -> "while"                {% nil %}
BREAK     -> "break"                {% nil %}