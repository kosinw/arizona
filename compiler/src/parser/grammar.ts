// Generated automatically by nearley, version 2.20.1
// http://github.com/Hardmath123/nearley
// Bypasses TS6133. Allow declared but unused functions.
// @ts-ignore
function id(d: any[]): any { return d[0]; }
declare var type: any;
declare var identifier: any;
declare var float: any;
declare var integer: any;
declare var string: any;
declare var char: any;
declare var bool: any;

import * as moo from 'moo';
import { makeLexer } from '.';
import makeSyntaxTypes, { flatten, nth, nil, compose, drop } from './syntax';

const lexer = makeLexer();
const syntax = makeSyntaxTypes();

interface NearleyToken {
  value: any;
  [key: string]: any;
};

interface NearleyLexer {
  reset: (chunk: string, info: any) => void;
  next: () => NearleyToken | undefined;
  save: () => any;
  formatError: (token: never) => string;
  has: (tokenType: string) => boolean;
};

interface NearleyRule {
  name: string;
  symbols: NearleySymbol[];
  postprocess?: (d: any[], loc?: number, reject?: {}) => any;
};

type NearleySymbol = string | { literal: any } | { test: (token: any) => boolean };

interface Grammar {
  Lexer: NearleyLexer | undefined;
  ParserRules: NearleyRule[];
  ParserStart: string;
};

const grammar: Grammar = {
  Lexer: lexer,
  ParserRules: [
    {"name": "_$ebnf$1", "symbols": []},
    {"name": "_$ebnf$1", "symbols": ["_$ebnf$1", "wschar"], "postprocess": (d) => d[0].concat([d[1]])},
    {"name": "_", "symbols": ["_$ebnf$1"], "postprocess": function(d) {return null;}},
    {"name": "__$ebnf$1", "symbols": ["wschar"]},
    {"name": "__$ebnf$1", "symbols": ["__$ebnf$1", "wschar"], "postprocess": (d) => d[0].concat([d[1]])},
    {"name": "__", "symbols": ["__$ebnf$1"], "postprocess": function(d) {return null;}},
    {"name": "wschar", "symbols": [/[ \t\n\v\f]/], "postprocess": id},
    {"name": "Statement", "symbols": ["ExpressionStatement"], "postprocess": id},
    {"name": "Statement", "symbols": ["AssignmentStatement"], "postprocess": id},
    {"name": "Statement", "symbols": ["ReturnStatement"], "postprocess": id},
    {"name": "ReturnStatement", "symbols": ["RETURN", "__", "Expression", "TERMINATOR"], "postprocess": syntax.return},
    {"name": "ReturnStatement", "symbols": ["RETURN", "TERMINATOR"], "postprocess": syntax.return},
    {"name": "AssignmentStatement", "symbols": ["_Assignment", "TERMINATOR"], "postprocess": id},
    {"name": "_Assignment", "symbols": ["Subscript", "_", {"literal":"="}, "_", "Expression"], "postprocess": syntax.assignment},
    {"name": "_Assignment", "symbols": ["Subscript", "_", {"literal":"+="}, "_", "Expression"], "postprocess": syntax.assignment},
    {"name": "_Assignment", "symbols": ["Subscript", "_", {"literal":"-="}, "_", "Expression"], "postprocess": syntax.assignment},
    {"name": "_Assignment", "symbols": ["Subscript", "_", {"literal":"/="}, "_", "Expression"], "postprocess": syntax.assignment},
    {"name": "_Assignment", "symbols": ["Subscript", "_", {"literal":"*="}, "_", "Expression"], "postprocess": syntax.assignment},
    {"name": "_Assignment", "symbols": ["Subscript", "_", {"literal":"%="}, "_", "Expression"], "postprocess": syntax.assignment},
    {"name": "ExpressionStatement", "symbols": ["Expression", "TERMINATOR"], "postprocess": id},
    {"name": "Expression", "symbols": ["Binary"], "postprocess": id},
    {"name": "Binary", "symbols": ["Logical"], "postprocess": id},
    {"name": "Logical", "symbols": ["Logical", "_", {"literal":"||"}, "_", "Bitwise"], "postprocess": syntax.binary},
    {"name": "Logical", "symbols": ["Logical", "_", {"literal":"&&"}, "_", "Bitwise"], "postprocess": syntax.binary},
    {"name": "Logical", "symbols": ["Bitwise"], "postprocess": id},
    {"name": "Bitwise", "symbols": ["Bitwise", "_", {"literal":"|"}, "_", "Sum"], "postprocess": syntax.binary},
    {"name": "Bitwise", "symbols": ["Bitwise", "_", {"literal":"^"}, "_", "Sum"], "postprocess": syntax.binary},
    {"name": "Bitwise", "symbols": ["Bitwise", "_", {"literal":"&"}, "_", "Sum"], "postprocess": syntax.binary},
    {"name": "Bitwise", "symbols": ["Equality"], "postprocess": id},
    {"name": "Equality", "symbols": ["Equality", "_", {"literal":"=="}, "_", "Comparison"], "postprocess": syntax.binary},
    {"name": "Equality", "symbols": ["Equality", "_", {"literal":"!="}, "_", "Comparison"], "postprocess": syntax.binary},
    {"name": "Equality", "symbols": ["Comparison"], "postprocess": id},
    {"name": "Comparison", "symbols": ["Comparison", "_", {"literal":"<"}, "_", "Shift"], "postprocess": syntax.binary},
    {"name": "Comparison", "symbols": ["Comparison", "_", {"literal":">"}, "_", "Shift"], "postprocess": syntax.binary},
    {"name": "Comparison", "symbols": ["Comparison", "_", {"literal":"<="}, "_", "Shift"], "postprocess": syntax.binary},
    {"name": "Comparison", "symbols": ["Comparison", "_", {"literal":">="}, "_", "Shift"], "postprocess": syntax.binary},
    {"name": "Comparison", "symbols": ["Shift"], "postprocess": id},
    {"name": "Shift", "symbols": ["Shift", "_", {"literal":">>"}, "_", "Sum"], "postprocess": syntax.binary},
    {"name": "Shift", "symbols": ["Shift", "_", {"literal":"<<"}, "_", "Sum"], "postprocess": syntax.binary},
    {"name": "Shift", "symbols": ["Sum"], "postprocess": id},
    {"name": "Sum", "symbols": ["Sum", "_", {"literal":"+"}, "_", "Product"], "postprocess": syntax.binary},
    {"name": "Sum", "symbols": ["Sum", "_", {"literal":"-"}, "_", "Product"], "postprocess": syntax.binary},
    {"name": "Sum", "symbols": ["Product"], "postprocess": id},
    {"name": "Product", "symbols": ["Product", "_", {"literal":"*"}, "_", "Unary"], "postprocess": syntax.binary},
    {"name": "Product", "symbols": ["Product", "_", {"literal":"/"}, "_", "Unary"], "postprocess": syntax.binary},
    {"name": "Product", "symbols": ["Product", "_", {"literal":"%"}, "_", "Unary"], "postprocess": syntax.binary},
    {"name": "Product", "symbols": ["Unary"], "postprocess": id},
    {"name": "Unary", "symbols": [{"literal":"!"}, "Call"], "postprocess": syntax.unary},
    {"name": "Unary", "symbols": [{"literal":"~"}, "Call"], "postprocess": syntax.unary},
    {"name": "Unary", "symbols": [{"literal":"!"}, "Call"], "postprocess": syntax.unary},
    {"name": "Unary", "symbols": [{"literal":"+"}, "Call"], "postprocess": syntax.unary},
    {"name": "Unary", "symbols": ["Call"], "postprocess": id},
    {"name": "Call", "symbols": ["Subscript", "_", "LP", "_", "ArgumentList", "_", "RP"], "postprocess": compose(syntax.call, flatten)},
    {"name": "Call", "symbols": ["Subscript", "_", "LP", "_", "RP"], "postprocess": syntax.call},
    {"name": "Call", "symbols": ["Subscript"], "postprocess": id},
    {"name": "ArgumentList", "symbols": ["Expression"], "postprocess": id},
    {"name": "ArgumentList", "symbols": ["NativeType"], "postprocess": id},
    {"name": "ArgumentList", "symbols": ["Expression", "_", "COMMA", "_", "ArgumentList"], "postprocess": flatten},
    {"name": "Subscript", "symbols": ["Access", "LSB", "_", "Expression", "_", "RSB", "Subscript"], "postprocess": syntax.subscript},
    {"name": "Subscript", "symbols": ["Access", "LSB", "_", "Expression", "_", "RSB"], "postprocess": syntax.subscript},
    {"name": "Subscript", "symbols": ["Access"], "postprocess": id},
    {"name": "Access", "symbols": ["Access", "DOT", "Identifier"], "postprocess": compose(syntax.access, drop)},
    {"name": "Access", "symbols": ["NativeType", "DOT", "Identifier"], "postprocess": compose(syntax.access, drop)},
    {"name": "Access", "symbols": ["Grouping"], "postprocess": id},
    {"name": "Grouping", "symbols": ["LP", "_", "Expression", "_", "RP"], "postprocess": nth(2)},
    {"name": "Grouping", "symbols": ["Atom"], "postprocess": id},
    {"name": "Atom", "symbols": ["Identifier"], "postprocess": id},
    {"name": "Atom", "symbols": ["StringLiteral"], "postprocess": id},
    {"name": "Atom", "symbols": ["CharLiteral"], "postprocess": id},
    {"name": "Atom", "symbols": ["FloatLiteral"], "postprocess": id},
    {"name": "Atom", "symbols": ["IntegerLiteral"], "postprocess": id},
    {"name": "NativeType", "symbols": [(lexer.has("type") ? {type: "type"} : type)], "postprocess": syntax.type},
    {"name": "Identifier", "symbols": [(lexer.has("identifier") ? {type: "identifier"} : identifier)], "postprocess": syntax.identifier},
    {"name": "FloatLiteral", "symbols": [(lexer.has("float") ? {type: "float"} : float)], "postprocess": syntax.float},
    {"name": "IntegerLiteral", "symbols": [(lexer.has("integer") ? {type: "integer"} : integer)], "postprocess": syntax.integer},
    {"name": "StringLiteral", "symbols": [(lexer.has("string") ? {type: "string"} : string)], "postprocess": syntax.string},
    {"name": "CharLiteral", "symbols": [(lexer.has("char") ? {type: "char"} : char)], "postprocess": syntax.char},
    {"name": "BoolLiteral", "symbols": [(lexer.has("bool") ? {type: "bool"} : bool)], "postprocess": syntax.bool},
    {"name": "LSB", "symbols": [{"literal":"["}], "postprocess": nil},
    {"name": "RSB", "symbols": [{"literal":"]"}], "postprocess": nil},
    {"name": "LP", "symbols": [{"literal":"("}], "postprocess": nil},
    {"name": "RANGLE", "symbols": [{"literal":">"}], "postprocess": nil},
    {"name": "RP", "symbols": [{"literal":")"}], "postprocess": nil},
    {"name": "DOT", "symbols": [{"literal":"."}], "postprocess": nil},
    {"name": "COMMA", "symbols": [{"literal":","}], "postprocess": nil},
    {"name": "LANGLE", "symbols": [{"literal":"<"}], "postprocess": nil},
    {"name": "TERMINATOR", "symbols": ["_", {"literal":";"}], "postprocess": nil},
    {"name": "FN", "symbols": [{"literal":"fn"}], "postprocess": nil},
    {"name": "LET", "symbols": [{"literal":"let"}], "postprocess": nil},
    {"name": "CONST", "symbols": [{"literal":"const"}], "postprocess": nil},
    {"name": "EXPORT", "symbols": [{"literal":"export"}], "postprocess": nil},
    {"name": "IMPORT", "symbols": [{"literal":"import"}], "postprocess": nil},
    {"name": "FROM", "symbols": [{"literal":"from"}], "postprocess": nil},
    {"name": "RETURN", "symbols": [{"literal":"return"}], "postprocess": nil},
    {"name": "TYPE", "symbols": [{"literal":"type"}], "postprocess": nil},
    {"name": "IF", "symbols": [{"literal":"if"}], "postprocess": nil},
    {"name": "ELSE", "symbols": [{"literal":"else"}], "postprocess": nil},
    {"name": "FOR", "symbols": [{"literal":"for"}], "postprocess": nil},
    {"name": "WHILE", "symbols": [{"literal":"while"}], "postprocess": nil},
    {"name": "BREAK", "symbols": [{"literal":"break"}], "postprocess": nil}
  ],
  ParserStart: "Statement",
};

export default grammar;
