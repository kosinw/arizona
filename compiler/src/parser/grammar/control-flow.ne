IfStatement ->
      IF _ Expression _ Block                  {% syntax.node(SyntaxType.IfStatement) %}
    | IF _ Expression _ Block _ ElseStatement  {% syntax.node(SyntaxType.IfStatement) %}

ElseStatement ->
      ELSE _ Block                             {% syntax.else %}
    | ELSE _ IfStatement                       {% syntax.else %}

ForStatement ->
      FOR _ _Assignment TERMINATOR _ Expression TERMINATOR _ _Assignment _ Block
                                              {% syntax.for %}
    | FOR _ LP _ _Assignment TERMINATOR _ Expression TERMINATOR _ _Assignment _ RP _ Block
                                              {% syntax.for %}
    | FOR _ VariableDeclaration _ Expression TERMINATOR _ _Assignment _ Block
                                              {% syntax.for %}
    | FOR _ LP _ VariableDeclaration _ Expression TERMINATOR _ _Assignment _ RP _ Block
                                              {% syntax.for %}

WhileStatement -> WHILE _ Expression _ Block  {% syntax.while %}

Break     -> BREAK TERMINATOR                  {% syntax.node(SyntaxType.BreakStatement) %}
Continue  -> CONTINUE TERMINATOR               {% syntax.node(SyntaxType.ContinueStatement) %}
