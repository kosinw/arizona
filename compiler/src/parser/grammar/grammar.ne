# Grammar for Arizona programming langauge

@preprocessor typescript
@builtin "whitespace.ne"

@include "./types.ne"
@include "./control-flow.ne"
@include "./use.ne"

@{%
import { lexer } from './lexer';
import { SyntaxType } from './types';
import makeSyntaxTypes, { extendNode, flatten, nth, nil, compose, drop } from './syntax';

const syntax = makeSyntaxTypes();
const ll = lexer();
%}

@lexer ll

Program ->
      _                                 {% compose(syntax.node(SyntaxType.RootNode, { value: 'root' }), flatten) %}
    | _ TopLevelDeclarations _          {% compose(syntax.node(SyntaxType.RootNode, { value: 'root' }), flatten) %}

TopLevelDeclarations ->
    TopLevelDeclaration                             {% flatten %}
    | TopLevelDeclaration _ TopLevelDeclarations    {% compose(drop, flatten, flatten) %}

TopLevelDeclaration ->
      FunctionDeclaration                       {% id %}
    | GlobalVariableDeclaration                 {% id %}
    | GlobalImmutableDeclaration                {% id %}
    | Export                                    {% id %}
    | UseFunctionDeclaration                    {% id %}

Export ->
      EXPORT __ GlobalImmutableDeclaration            {% syntax.node(SyntaxType.Export, { value: 'export' }) %}
    | EXPORT __ FunctionDeclaration                   {% syntax.node(SyntaxType.Export, { value: 'export' }) %}
    | EXPORT __ GlobalVariableDeclaration             {% syntax.node(SyntaxType.Export, { value: 'export' }) %}

Block ->
      LCB _ RCB                                   {% syntax.node(SyntaxType.Block) %}
    | LCB _ Statements _ RCB                      {% compose(syntax.node(SyntaxType.Block), flatten) %}

Statements ->
      Statement                                                         {% drop %}
    | Statement _ Statements                                            {% flatten %}

FunctionDeclaration ->
      FN __ Identifier _ FunctionParameters  _ Block                    {% syntax.voidfn %}
    | FN __ Identifier _ FunctionParameters _ FunctionResult _ Block    {% syntax.fn %}

FunctionHeaderDeclaration ->
      FN __ Identifier _ TypeParameters                    {% syntax.fnheader %}
    | FN __ Identifier _ TypeParameters _ FunctionResult   {% syntax.fnheader %}

FunctionResult ->
    RARROW _ Type                                           {% compose(syntax.result, drop) %}

FunctionParameters ->
      LP _ RP                                               {% syntax.node(SyntaxType.FunctionParameters) %}
    | LP _ Parameters _ RP                                  {% compose(syntax.node(SyntaxType.FunctionParameters), flatten, flatten) %}

TypeParameters ->
      LP _ RP                                               {% syntax.node(SyntaxType.FunctionParameters) %}
    | LP _ Types _ RP                                       {% compose(syntax.node(SyntaxType.FunctionParameters), flatten, flatten) %}

Parameters ->
      NameAndType                                           {% id %}
    | NameAndType _ COMMA _ Parameters                      {% flatten %}

Types ->
      Type                                           {% id %}
    | Type _ COMMA _ Types                           {% flatten %}

AnonymousBlock ->
    Block                                                 {% id %}

Statement ->
      ExpressionStatement                                 {% id %}
    | AssignmentStatement                                 {% id %}
    | ReturnStatement                                     {% id %}
    | VariableDeclaration                                 {% id %}
    | ImmutableDeclaration                                {% id %}
    | IfStatement                                         {% id %}
    | AnonymousBlock                                      {% id %}
    | ForStatement                                        {% id %}
    | WhileStatement                                      {% id %}
    | Break                                               {% id %}
    | Continue                                            {% id %}

GlobalVariableDeclaration ->
      LET _ NameAndType _ EQUALS _ Atom TERMINATOR              {% syntax.declaration(SyntaxType.GlobalVariableDeclaration) %}
    | LET _ NameAndType TERMINATOR                              {% syntax.declaration(SyntaxType.GlobalVariableDeclaration) %}

GlobalImmutableDeclaration ->
      CONST _ NameAndType _ EQUALS _ Atom TERMINATOR            {% syntax.declaration(SyntaxType.GlobalImmutableDeclaration) %}

VariableDeclaration ->
      LET _ NameAndType _ EQUALS _ Expression TERMINATOR        {% syntax.declaration(SyntaxType.VariableDeclaration) %}
    | LET _ NameAndType TERMINATOR                              {% syntax.declaration(SyntaxType.VariableDeclaration) %}

ImmutableDeclaration ->
      CONST _ NameAndType _ EQUALS _ Expression TERMINATOR      {% syntax.declaration(SyntaxType.ImmutableDeclaration) %}

NameAndType -> Identifier _ COLON _ DeclType                    {% syntax.node(SyntaxType.Pair) %}

DeclType -> Type                                                {% compose(extendNode({ type: SyntaxType.DeclType }), nth(0)) %}

ReturnStatement ->
      RETURN __ Expression TERMINATOR                           {% syntax.return %}
    | RETURN TERMINATOR                                         {% syntax.return %}

AssignmentStatement -> _Assignment TERMINATOR                   {% id %}

_Assignment ->
      Subscript _ "="  _ Expression                     {% syntax.assignment %}
    | Subscript _ "+=" _ Expression                     {% syntax.assignment %}
    | Subscript _ "-=" _ Expression                     {% syntax.assignment %}
    | Subscript _ "/=" _ Expression                     {% syntax.assignment %}
    | Subscript _ "*=" _ Expression                     {% syntax.assignment %}
    | Subscript _ "%=" _ Expression                     {% syntax.assignment %}

ExpressionStatement -> Expression TERMINATOR          {% id %}

Expression -> Binary                                    {% id %}

Binary -> Logical                   {% id %}

Logical ->
      Logical _ "||" _ Bitwise        {% syntax.binary %}
    | Logical _ "&&" _ Bitwise        {% syntax.binary %}
    | Bitwise                         {% id %}

Bitwise ->
      Bitwise _ "|" _ Sum             {% syntax.binary %}
    | Bitwise _ "^" _ Sum             {% syntax.binary %}
    | Bitwise _ "&" _ Sum             {% syntax.binary %}
    | Equality                        {% id %}

Equality ->
      Equality _ "==" _ Comparison      {% syntax.binary %}
    | Equality _ "!=" _ Comparison      {% syntax.binary %}
    | Comparison                        {% id %}

Comparison ->
      Comparison _ "<" _ Shift        {% syntax.binary %}
    | Comparison _ ">" _ Shift        {% syntax.binary %}
    | Comparison _ "<=" _ Shift       {% syntax.binary %}
    | Comparison _ ">=" _ Shift       {% syntax.binary %}
    | Shift                           {% id %}

Shift ->
      Shift _ ">>" _ Sum            {% syntax.binary %}
    | Shift _ "<<" _ Sum            {% syntax.binary %}
    | Sum                           {% id %}

Sum ->
      Sum _ "+" _ Product                       {% syntax.binary %}
    | Sum _ "-" _ Product                       {% syntax.binary %}
    | Product                                   {% id %}

Product ->
      Product _ "*" _ Unary                     {% syntax.binary %}
    | Product _ "/" _ Unary                     {% syntax.binary %}
    | Product _ "%" _ Unary                     {% syntax.binary %}
    | Unary                                     {% id %}

Unary ->
      "!" Call                                  {% syntax.unary %}
    | "~" Call                                  {% syntax.unary %}
    | "!" Call                                  {% syntax.unary %}
    | "+" Call                                  {% syntax.unary %}
    | Call                                      {% id %}

Call ->
      Subscript _ LP _ ArgumentList _ RP        {% compose(syntax.call, flatten) %}
    | Subscript _ LP _ RP                       {% syntax.call %}
    | Subscript                                 {% id %}

ArgumentList ->
      Expression                                {% id %}
    | NativeType                                {% id %}
    | Expression _ COMMA _ ArgumentList         {% flatten %}

Subscript ->
      Access LSB _ Expression _ RSB Subscript   {% syntax.subscript %}
    | Access LSB _ Expression _ RSB             {% syntax.subscript %}
    | Access                                    {% id %}

Access ->
      Access DOT Identifier         {% compose(syntax.access, drop) %}
    | NativeType DOT Identifier     {% compose(syntax.access, drop) %}
    | Grouping                      {% id %}

Grouping ->
      LP _ Expression _ RP            {% nth(2) %}
    | Atom                          {% id %}

Atom ->
      Identifier                {% id %}
    | StringLiteral             {% id %}
    | CharLiteral               {% id %}
    | FloatLiteral              {% id %}
    | IntegerLiteral            {% id %}
    | BoolLiteral               {% id %}

NativeType      -> %type            {% syntax.type %}
Identifier      -> %identifier      {% syntax.identifier %}
FloatLiteral    -> %float           {% syntax.float %}
IntegerLiteral  -> %integer         {% syntax.integer %}
StringLiteral   -> %string          {% syntax.string %}
CharLiteral     -> %char            {% syntax.char %}
BoolLiteral     -> %bool            {% syntax.bool %}

# Punctuators
LSB         -> "["     {% nil %}
RSB         -> "]"     {% nil %}
LP          -> "("     {% nil %}
RP          -> ")"     {% nil %}
LCB         -> "{"     {% nil %}
RCB         -> "}"     {% nil %}
RANGLE      -> ">"     {% nil %}
DOT         -> "."     {% nil %}
COMMA       -> ","     {% nil %}
LANGLE      -> "<"     {% nil %}
TERMINATOR  -> _ ";"   {% nil %}
EQUALS      -> "="     {% nil %}
COLON       -> ":"     {% nil %}
RARROW      -> "->"    {% nil %}

# Keywords
FN        -> "fn"                   {% nil %}
VOID      -> "void"                 {% nil %}
LET       -> "let"                  {% nil %}
CONST     -> "const"                {% nil %}
EXPORT    -> "export"               {% nil %}
USE       -> "use"                  {% nil %}
FROM      -> "from"                 {% nil %}
RETURN    -> "return"               {% nil %}
TYPE      -> "type"                 {% nil %}
IF        -> "if"                   {% nil %}
ELSE      -> "else"                 {% nil %}
FOR       -> "for"                  {% nil %}
WHILE     -> "while"                {% nil %}
BREAK     -> "break"                {% nil %}
CONTINUE  -> "continue"             {% nil %}