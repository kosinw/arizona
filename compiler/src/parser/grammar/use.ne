UseFunctionDeclaration ->
    USE __ FunctionHeaderDeclaration _ FROM _ StringLiteral TERMINATOR  {% syntax.node(SyntaxType.UseFunctionDeclaration) %}
