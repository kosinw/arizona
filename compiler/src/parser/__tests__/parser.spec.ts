import { describe, it, expect } from '@jest/globals';

import grammar from '../grammar';
import * as util from 'util';
import dedent from 'dedent';
import { Grammar, Parser } from 'nearley';

const depthless = (d: any) => util.inspect(d, { depth: null });

describe('parser tests', () => {
  it('can parse a very complex program', () => {
    const src = dedent`
      use fn setPixel(i32, i32) -> f32 from "env";

      let i: i32 = 15;

      export fn echo(a: i32, c: f64) -> i32 {
        while true {
          let j: i32 = 16;
          let k: f64 = 1.23456;
          let l: i32 = 0x3a;
          continue;
          break;
        }

        return (15 ^ 18) << 3 * 99;
      }
    `;
    
    const parser = new Parser(Grammar.fromCompiled(grammar));
    parser.feed(src);
    const ast = parser.results[0];

    // expect(parser.results.length).toEqual(1);
    expect(ast).toMatchSnapshot();
  });
});
