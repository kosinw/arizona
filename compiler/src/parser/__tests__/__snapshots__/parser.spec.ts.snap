// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser tests can parse a very complex program 1`] = `
Object {
  "params": Array [
    Object {
      "params": Array [
        Object {
          "params": Array [
            Object {
              "params": Array [],
              "staticType": null,
              "type": "Identifier",
              "value": "setPixel",
            },
            Object {
              "params": Array [
                Object {
                  "params": Array [],
                  "staticType": "i32",
                  "type": "Type",
                  "value": "i32",
                },
                Object {
                  "params": Array [],
                  "staticType": "i32",
                  "type": "Type",
                  "value": "i32",
                },
              ],
              "staticType": null,
              "type": "FunctionParameters",
              "value": "",
            },
            Object {
              "params": Array [
                Object {
                  "params": Array [],
                  "staticType": "f32",
                  "type": "Type",
                  "value": "f32",
                },
              ],
              "staticType": "f32",
              "type": "FunctionResult",
              "value": "",
            },
          ],
          "staticType": null,
          "type": "FunctionHeaderDeclaration",
          "value": "setPixel",
        },
        Object {
          "params": Array [],
          "staticType": "i32",
          "type": "StringLiteral",
          "value": "env",
        },
      ],
      "staticType": null,
      "type": "UseFunctionDeclaration",
      "value": "",
    },
    Object {
      "params": Array [
        Object {
          "params": Array [],
          "staticType": "i32",
          "type": "IntegerLiteral",
          "value": 15,
        },
      ],
      "staticType": "i32",
      "type": "GlobalVariableDeclaration",
      "value": "i",
    },
    Object {
      "params": Array [
        Object {
          "params": Array [
            Object {
              "params": Array [],
              "staticType": null,
              "type": "Identifier",
              "value": "echo",
            },
            Object {
              "params": Array [
                Object {
                  "params": Array [
                    Object {
                      "params": Array [],
                      "staticType": null,
                      "type": "Identifier",
                      "value": "a",
                    },
                    Object {
                      "params": Array [],
                      "staticType": "i32",
                      "type": "DeclType",
                      "value": "i32",
                    },
                  ],
                  "staticType": null,
                  "type": "Pair",
                  "value": "",
                },
                Object {
                  "params": Array [
                    Object {
                      "params": Array [],
                      "staticType": null,
                      "type": "Identifier",
                      "value": "c",
                    },
                    Object {
                      "params": Array [],
                      "staticType": "f64",
                      "type": "DeclType",
                      "value": "f64",
                    },
                  ],
                  "staticType": null,
                  "type": "Pair",
                  "value": "",
                },
              ],
              "staticType": null,
              "type": "FunctionParameters",
              "value": "",
            },
            Object {
              "params": Array [
                Object {
                  "params": Array [],
                  "staticType": "i32",
                  "type": "Type",
                  "value": "i32",
                },
              ],
              "staticType": "i32",
              "type": "FunctionResult",
              "value": "",
            },
            Object {
              "params": Array [
                Object {
                  "params": Array [
                    Object {
                      "params": Array [],
                      "staticType": null,
                      "type": "Noop",
                      "value": "",
                    },
                    Object {
                      "params": Array [],
                      "staticType": "bool",
                      "type": "BoolLiteral",
                      "value": true,
                    },
                    Object {
                      "params": Array [
                        Object {
                          "params": Array [
                            Object {
                              "params": Array [],
                              "staticType": "i32",
                              "type": "IntegerLiteral",
                              "value": 16,
                            },
                          ],
                          "staticType": "i32",
                          "type": "VariableDeclaration",
                          "value": "j",
                        },
                        Object {
                          "params": Array [
                            Object {
                              "params": Array [],
                              "staticType": "f32",
                              "type": "FloatLiteral",
                              "value": 1.23456,
                            },
                          ],
                          "staticType": "f64",
                          "type": "VariableDeclaration",
                          "value": "k",
                        },
                        Object {
                          "params": Array [
                            Object {
                              "params": Array [],
                              "staticType": "i32",
                              "type": "IntegerLiteral",
                              "value": 58,
                            },
                          ],
                          "staticType": "i32",
                          "type": "VariableDeclaration",
                          "value": "l",
                        },
                        Object {
                          "params": Array [],
                          "staticType": null,
                          "type": "ContinueStatement",
                          "value": "",
                        },
                        Object {
                          "params": Array [],
                          "staticType": null,
                          "type": "BreakStatement",
                          "value": "",
                        },
                      ],
                      "staticType": null,
                      "type": "Block",
                      "value": "",
                    },
                  ],
                  "staticType": null,
                  "type": "LoopStatement",
                  "value": "",
                },
                Object {
                  "params": Array [
                    Object {
                      "params": Array [
                        Object {
                          "params": Array [
                            Object {
                              "params": Array [],
                              "staticType": "i32",
                              "type": "IntegerLiteral",
                              "value": 15,
                            },
                            Object {
                              "params": Array [],
                              "staticType": "i32",
                              "type": "IntegerLiteral",
                              "value": 18,
                            },
                          ],
                          "staticType": null,
                          "type": "BinaryExpression",
                          "value": "^",
                        },
                        Object {
                          "params": Array [
                            Object {
                              "params": Array [],
                              "staticType": "i32",
                              "type": "IntegerLiteral",
                              "value": 3,
                            },
                            Object {
                              "params": Array [],
                              "staticType": "i32",
                              "type": "IntegerLiteral",
                              "value": 99,
                            },
                          ],
                          "staticType": null,
                          "type": "BinaryExpression",
                          "value": "*",
                        },
                      ],
                      "staticType": null,
                      "type": "BinaryExpression",
                      "value": "<<",
                    },
                  ],
                  "staticType": null,
                  "type": "ReturnStatement",
                  "value": "",
                },
              ],
              "staticType": null,
              "type": "Block",
              "value": "",
            },
          ],
          "staticType": null,
          "type": "FunctionDeclaration",
          "value": "echo",
        },
      ],
      "staticType": null,
      "type": "Export",
      "value": "export",
    },
  ],
  "staticType": null,
  "type": "RootNode",
  "value": "root",
}
`;
